Back navigation to previous tab

cmonfortep opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.
-->

Task/Issue URL: https://app.asana.com/0/414730916066338/1187653972293689/f
Tech Design URL: 
CC: 

**Description**:
Tries to fix the following issue:

Current behavior:
- User clicks on a link from tab A
- New tab is opened (Tab B)
- User presses back
- Home page shown (in Tab B)
- User presses back again (in Tab B)
- App is closed

Expected behavior:
- User clicks on a link from tab A
- New tab is opened (Tab B)
- User presses back (Tab B)
- Tab B closed
- Tab A opened

**Related open issues:**
https://github.com/duckduckgo/Android/issues/815
https://github.com/duckduckgo/Android/issues/696

**Steps to test this PR**:

Test: Open link on a new tab
1. Open the app
1. Perform a search or visit any website (Tab A)
1. long press on a link
1. open on a new tab (Tab B)
1. ensure new tab is opened with expected link
1. press back
1. ensure Tab B is closed, and user goes back to Tab A

Test: Open link on a new tab (source tab deleted)
1. Open the app
1. Perform a search or visit any website (Tab A)
1. long press on a link
1. open on a new tab (Tab B)
1. ensure new tab is opened with expected link
1. Go to tab switcher, delete Tab A
1. Go to Tab B
1. Press back
1. Ensure the home screen is shown

Test: Open link on a background tab
1. Open the app
1. Perform a search or visit any website (Tab A)
1. long press on a link
1. Click on "Open in Background Tab" (Tab B)
1. ensure new tab is created (tabs counter)
1. Go to that tab (open tabswitcher, and change tab)
1. press back
1. ensure Tab B is closed, and user goes back to Tab A

Test: Open link on a background tab (source tab deleted)
1. Open the app
1. Perform a search or visit any website (Tab A)
1. long press on a link
1. Click on "Open in Background Tab" (Tab B)
1. ensure new tab is created (tabs counter)
1. Go to tab switcher, delete Tab A
1. Go to Tab B
1. Press back
1. Ensure the home screen is shown

Test: Open link in new tab (target=_blank)
1. Open the app
1. Visit https://www.w3schools.com/code/tryit.asp?filename=GGETHXPDSEMN (Tab A)
1. Press run and click on "test" or "link in new tab"
1. ensure new tab is opened (Tab B)
1. press back
1. ensure Tab B is closed, and user goes back to Tab A

Test: Open link in new tab (target=_blank / source tab deleted)
1. Open the app
1. Visit https://www.w3schools.com/code/tryit.asp?filename=GGETHXPDSEMN (Tab A)
1. Press run and click on "test" or "link in new tab"
1. ensure new tab is opened (Tab B)
1. Go to tab switcher, delete Tab A
1. Go to Tab B
1. Press back
1. Ensure the home screen is shown

**Steps to test previous logic is preserved**:

Test: Open a New Empty tab
1. Open the app
1. create a new tab manually from overflow menu or tabswitcher
1. Ensure new tab is created
1. perform a search o visit a website
1. press back
1. Ensure home screen is shown

Test: External links
1. Go to another app where you have a link to click
1. Open that link with DDG
1. Ensure link is opened in a new tab
1. Press back
1. Ensure previous app is opened again
1. Open DDG app again via task manager
1. Press back on the tab you previously opened
1. Ensure App gets closed

Test: Search widget
1. Click on our search widget
1. search something or click any suggestion
1. Ensure new tab is created with your search term
1. Press back
1. Ensure App gets closed

Test: Shortcut
1. Create a shortcut to any website
1. Click on the shortcut
1. Ensure new tab is created with your search term
1. Press back
1. Ensure App gets closed

**Test database migration**:
1. Install current `main` branch
1. Open as many tabs as you want
1. Install this branch
1. Open app again
1. Ensure app doesn't crash

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

cmonfortep commented on app/src/main/java/com/duckduckgo/app/global/db/AppDatabase.kt:
```diff
@@ -318,7 +318,22 @@ class MigrationsProvider(
     val MIGRATION_23_TO_24: Migration = object : Migration(23, 24) {
         override fun migrate(database: SupportSQLiteDatabase) {
-            database.execSQL("ALTER TABLE `tabs` ADD COLUMN `sourceTabId` TEXT")
+            //https://stackoverflow.com/a/57797179/980345
```
@CDRussell When testing migrations I detected an issue trying to `ALTER` a table to introduce a `FOREIGN` key. It seems it's not supported and the workaround is to migrate -> drop -> create a new table.

I have my worries (based on nothing in concrete, just a feeling), don't know if this can be an issue for users with many open tabs...

cmonfortep commented on app/src/main/java/com/duckduckgo/app/browser/WebViewClientListener.kt:
```diff
@@ -42,7 +42,6 @@ interface WebViewClientListener {
     fun exitFullScreen()
     fun showFileChooser(filePathCallback: ValueCallback<Array<Uri>>, fileChooserParams: WebChromeClient.FileChooserParams)
     fun externalAppLinkClicked(appLink: SpecialUrlDetector.UrlType.IntentType)
-    fun openInNewTab(url: String?)
```
Removed this method, it was unused.

cmonfortep commented on app/src/androidTest/java/com/duckduckgo/app/browser/BrowserTabViewModelTest.kt:
```diff
@@ -1319,18 +1327,6 @@ class BrowserTabViewModelTest {
     }
 
-    @Test
-    fun whenOpenInNewTabThenOpenInNewTabCommandWithCorrectUrlSent() {
```
Method has been removed

cmonfortep commented on app/src/main/java/com/duckduckgo/app/tabs/model/TabDataRepository.kt:
```diff
@@ -83,7 +91,7 @@ class TabDataRepository @Inject constructor(
         return data
     }
 
-    override suspend fun add(tabId: String, data: MutableLiveData<Site>, skipHome: Boolean, isDefaultTab: Boolean, sourceTabId: String?) {
```
Made this method `private` since it was only used internally in this class.

cmonfortep commented on app/src/main/java/com/duckduckgo/app/tabs/model/TabRepository.kt:
```diff
@@ -29,11 +29,11 @@ interface TabRepository {
     /**
      * @return tabId of new record
      */
-    suspend fun add(url: String? = null, skipHome: Boolean = false, isDefaultTab: Boolean = false): String
```
I tried to simplify the overhead of thinking about `isDefaultTab` and in the code using these methods. `isDefaultTab` was `true` only in one place out of many different ones.

CDRussell replied:
👍

CDRussell commented on app/src/main/java/com/duckduckgo/app/browser/BrowserViewModel.kt:
```diff
@@ -126,22 +126,33 @@ class BrowserViewModel(
 
     suspend fun onTabsUpdated(tabs: List<TabEntity>?) {
-        if (tabs == null || tabs.isEmpty()) {
+        if (tabs.isNullOrEmpty()) {
```
nice tidy

CDRussell commented on app/src/main/java/com/duckduckgo/app/tabs/model/TabDataRepository.kt:
```diff
@@ -49,29 +47,39 @@ class TabDataRepository @Inject constructor(
     }
 
-    override suspend fun addWithSource(url: String?, skipHome: Boolean, isDefaultTab: Boolean): String {
+    override suspend fun addFromSourceTab(url: String?, skipHome: Boolean, sourceTabId: String?): String {
```
Could we make `sourceTabId: String?` non-nullable here?

CDRussell commented on app/src/main/java/com/duckduckgo/app/tabs/model/TabEntitiy.kt:
```diff
@@ -17,11 +17,20 @@
 @Entity(
     tableName = "tabs",
+        foreignKeys = [
+            ForeignKey(
```
thanks for looking into using foreign keys; think it worked out really neat

cmonfortep commented:
@CDRussell sorry for not pointing out how to test the database migration, I've added the info in the description (also in this comment)

The current production release has database version `23`, develop has version `24` (when we introduced `sourceTabId`). I wanted to avoid another database migration to just set `sourceTabId` as `Foreign key`. That's why I've just updated the migration step without changing the database version. Could you instead use `main` to validate the migration? 

**Test database migration**:
1. Install current `main` branch
1. Open as many tabs as you want
1. Install this branch
1. Open app again
1. Ensure app doesn't crash

CDRussell commented:
>Could you instead use main to validate the migration

I can do that. Though going forward i'd like to have nightly builds from `develop`, so once something is in `develop` we'd want to migrate any changes.
